internal/handlers/pg/cmd_query.go

```diff
@@ -16,47 +16,71 @@ package pg
 	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
 	"github.com/FerretDB/FerretDB/internal/util/must"
 	"github.com/FerretDB/FerretDB/internal/wire"
 )
 
 // CmdQuery implements HandlerInterface.
 func (h *Handler) CmdQuery(ctx context.Context, query *wire.OpQuery) (*wire.OpReply, error) {
-	if query.FullCollectionName == "admin.$cmd" {
-		switch cmd := query.Query.Command(); cmd {
-		case "ismaster", "isMaster": // both are valid
-			reply := &wire.OpReply{
-				NumberReturned: 1,
-				Documents: []*types.Document{must.NotFail(types.NewDocument(
-					"ismaster", true, // only lowercase
-					// topologyVersion
-					"maxBsonObjectSize", int32(types.MaxDocumentLen),
-					"maxMessageSizeBytes", int32(wire.MaxMsgLen),
-					"maxWriteBatchSize", int32(100000),
-					"localTime", time.Now(),
-					// logicalSessionTimeoutMinutes
-					"connectionId", int32(42),
-					"minWireVersion", common.MinWireVersion,
-					"maxWireVersion", common.MaxWireVersion,
-					"readOnly", false,
-					"ok", float64(1),
-				))},
+	cmd := query.Query.Command()
+	collection := query.FullCollectionName
+
+	// both are valid
+	if (cmd == "ismaster" || cmd == "isMaster") && collection == "admin.$cmd" {
+		return common.IsMaster(ctx, query)
+	}
+
+	// defaults to the database name if supplied on the connection string or $external
+	if cmd == "saslStart" &&
+		(collection == "$external.$cmd" || strings.HasSuffix(collection, "$cmd")) {
+		var emptyPayload types.Binary
+
+		if err := common.SASLStart(ctx, query.Query); err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		if _, err := h.DBPool(ctx); err != nil {
+			var pgErr *pgconn.PgError
+			if errors.As(err, &pgErr) && pgerrcode.IsInvalidAuthorizationSpecification(pgErr.Code) {
+				msg := "FerretDB failed to authenticate you in PostgreSQL:\n" +
+					strings.TrimSpace(pgErr.Error()) + "\n" +
+					"See https://docs.ferretdb.io/security/authentication/ for more details."
+
+				return nil, commonerrors.NewCommandErrorMsgWithArgument(
+					commonerrors.ErrAuthenticationFailed,
+					msg,
+					"payload",
+				)
 			}
-			return reply, nil
 
-		default:
-			msg := fmt.Sprintf("CmdQuery: unhandled command %q", cmd)
-			return nil, commonerrors.NewCommandErrorMsg(commonerrors.ErrNotImplemented, msg)
+			return nil, lazyerrors.Error(err)
 		}
-	}
 
-	msg := fmt.Sprintf("CmdQuery: unhandled collection %q", query.FullCollectionName)
+		return &wire.OpReply{
+			NumberReturned: 1,
+			Documents: []*types.Document{must.NotFail(types.NewDocument(
+				"conversationId", int32(1),
+				"done", true,
+				"payload", emptyPayload,
+				"ok", float64(1),
+			))},
+		}, nil
+	}
 
-	return nil, commonerrors.NewCommandErrorMsg(commonerrors.ErrNotImplemented, msg)
+	return nil, commonerrors.NewCommandErrorMsgWithArgument(
+		commonerrors.ErrNotImplemented,
+		fmt.Sprintf("CmdQuery: unhandled command %q for collection %q", cmd, collection),
+		"OpQuery: "+cmd,
```

Comment:
I think `cmd` is sufficient for telemetry here. I don't think extra constant `OpQuery: ` is necessary :hugs:

Reply:
I see, so `msg_saslstart.go` and `msg_query.go` should differentiate telemetry tag right. Do we want L63 to include OpQuery specific prefix, and vice versa for `msg_saslstart.go`? https://github.com/FerretDB/FerretDB/pull/3012/files#diff-33d418298c97b13391bb4b591add2cfdc0203a9f6f245277bfc034b7fec6e10cR63

Reply:
yeah, I think it should :)